cli/azd/pkg/project/framework_service_docker.go

```diff
@@ -297,6 +298,28 @@ func (p *dockerProject) packBuild(
 		userDefinedImage = true
 	}
 
+	if !userDefinedImage {
+		// Always default to port 80 for consistency across languages
+		environ = append(environ, "ORYX_RUNTIME_PORT=80")
+
+		if svc.OutputPath != "" && (svc.Language == ServiceLanguageTypeScript || svc.Language == ServiceLanguageJavaScript) {
+			inDockerOutputPath := path.Join("/workspace", svc.OutputPath)
+			// A dist folder has been set.
+			// We assume that the service is a front-end service, configuring a nginx web server to serve the static content
+			// produced.
+			environ = append(environ,
+				"ORYX_RUNTIME_IMAGE=nginx:1.25.2-bookworm",
+				fmt.Sprintf(
+					//nolint:lll
+					"ORYX_RUNTIME_SCRIPT=[ -d \"%s\" ] || { echo \"error: directory '%s' does not exist. ensure the 'dist' path in azure.yaml is specified correctly.\"; exit 1; } && "+
+						"rm -rf /usr/share/nginx/html && ln -sT %s /usr/share/nginx/html && "+
+						"nginx -g 'daemon off;'",
+					inDockerOutputPath,
+					svc.OutputPath,
+					inDockerOutputPath))
+		}
+	}
```

Comment:
This doesn't feel like the correct place for this script.  This component is for general purpose building of building docker based projects and now we're introducing edge case / service & language specific use cases in here.

Lets think on this a bit in how we could design a more extensible means to achieving this.

Reply:
A framework_service_docker is an azd-application construct that builds projects with `lang = docker`. With a Dockerfile, it is indeed correct that azd would treat docker as a blackbox. However, without a Dockerfile, azd may need to interrogate and integrate on the language level, based on user configuration from azure.yaml.

From the real-world app design perspective, I see the code organization inline with the end-to-end feature.

If we did introduce a new type `lang=docker-from-source`, you are right that this should not belong under framework_service_docker. This design can and will emerge overtime if users believe this is a better model. But organizing our code to reflect internal boundaries that do not represent the user model creates extensibility and separation as you pointed out with added complexity. Such complexity builds up over time, and we lose flexibility in the process of adapting to the ever-changing user requirements landscape.

If you're asking to simply move this out as a static helper, we can certainly do that. I don't see this being reused in different constructs, but if we like to separate things out just to lower the mental cost of tracking, that'd be fine.